test/slam_manifold_preintegration_update_test.cc

```diff
@@ -0,0 +1,303 @@
+/* ----------------------------------------------------------------------------
+ * SymForce - Copyright 2022, Skydio, Inc.
+ * This source code is under the Apache 2.0 license found in the LICENSE file.
+ * ---------------------------------------------------------------------------- */
+
+#include <numeric>
+#include <random>
+#include <thread>
+
+#include <Eigen/Dense>
+#include <catch2/catch_test_macros.hpp>
+
+#include <sym/factors/internal/imu_manifold_preintegration_update.h>
+#include <sym/rot3.h>
+#include <sym/util/epsilon.h>
+#include <symforce/slam/imu_preintegration/imu_preintegrator.h>
+
+using Eigen::Ref;
+using Eigen::Vector3d;
+using sym::Rot3d;
+
+/**
+ * Calculates what ImuManifoldPreintegrationUpdate should for the changes in orientation,
+ * velocity, and position.
+ */
+void UpdateState(const Rot3d& DR, const Ref<const Vector3d>& Dv, const Ref<const Vector3d>& Dp,
+                 const Ref<const Vector3d>& gyro, const Ref<const Vector3d>& accel,
+                 const Ref<const Vector3d>& gyro_bias, const Ref<const Vector3d>& accel_bias,
+                 const double dt, const double epsilon, Rot3d& new_DR, Ref<Vector3d> new_Dv,
+                 Ref<Vector3d> new_Dp) {
+  const auto corrected_accel = accel - accel_bias;
+  new_DR = DR * Rot3d::FromTangent((gyro - gyro_bias) * dt, epsilon);
+  new_Dv = Dv + DR * corrected_accel * dt;
+  new_Dp = Dp + Dv * dt + DR * corrected_accel * (0.5 * dt * dt);
+}
+
+TEST_CASE("Test ImuManifoldPreintegrationUpdate has correct DR, Dv, & Dp", "[slam]") {
+  std::mt19937 gen(1804);
+  std::normal_distribution<double> dist{0.0, 10.0};
+  const auto randomM3 = Eigen::Matrix3d::NullaryExpr([&]() { return dist(gen); });
+  const auto randomV3 = Eigen::Vector3d::NullaryExpr([&]() { return dist(gen); });
+  const auto randomM9 = Eigen::Matrix<double, 9, 9>::NullaryExpr([&]() { return dist(gen); });
+
+  for (int i_ = 0; i_ < 10; i_++) {
+    const sym::Rot3d DR = sym::Rot3d::Random(gen);
+    const Eigen::Vector3d Dv = randomV3;
+    const Eigen::Vector3d Dp = randomV3;
+    const Eigen::Matrix<double, 9, 9> covariance = randomM9;
+    const Eigen::Matrix3d DR_D_gyro_bias = randomM3;
+    const Eigen::Matrix3d Dv_D_gyro_bias = randomM3;
+    const Eigen::Matrix3d Dv_D_accel_bias = randomM3;
+    const Eigen::Matrix3d Dp_D_gyro_bias = randomM3;
+    const Eigen::Matrix3d Dp_D_accel_bias = randomM3;
```

Comment:
Hmm. I suppose I wanted to test that the values being tested don't depend on these other values. So I don't think I'd want to use 0. Using `NaN` does seem like an interesting way of achieving that though...

I think I'll go with that. Especially because that way it'll be easier to tell at a glance which inputs should and shouldn't affect the output.